packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/synthesize-resources.ts

```diff
@@ -0,0 +1,340 @@
+
+    // determine permissions needed for each trigger module
+    coreAnswers.permissions = await context.amplify.getTriggerPermissions(context, coreAnswers.triggers, 'auth', coreAnswers.resourceName);
+  } else if (previouslySaved) {
+    const targetDir = context.amplify.pathManager.getBackendDirPath();
+    Object.keys(previouslySaved).forEach(p => {
+      delete coreAnswers[p];
+    });
+    await context.amplify.deleteAllTriggers(previouslySaved, coreAnswers.resourceName, targetDir, context);
+  }
+  // remove unused coreAnswers.triggers key
+  if (coreAnswers.triggers && coreAnswers.triggers === '[]') {
+    delete coreAnswers.triggers;
+  }
+
+  // handle dependsOn data
+  const dependsOnKeys = Object.keys(triggerKeyValues).map(i => `${coreAnswers.resourceName}${i}`);
+  coreAnswers.dependsOn = context.amplify.dependsOnBlock(context, dependsOnKeys, 'Cognito');
+};
+
+export const createUserPoolGroups = async (context: any, resourceName: any, userPoolGroupList: any) => {
+  if (userPoolGroupList && userPoolGroupList.length > 0) {
+    const userPoolGroupPrecedenceList = [];
+
+    for (let i = 0; i < userPoolGroupList.length; i += 1) {
+      userPoolGroupPrecedenceList.push({
+        groupName: userPoolGroupList[i],
+        precedence: i + 1,
+      });
+    }
+
+    const userPoolGroupFile = path.join(
+      context.amplify.pathManager.getBackendDirPath(),
+      'auth',
+      'userPoolGroups',
+      'user-pool-group-precedence.json',
+    );
+
+    const userPoolGroupParams = path.join(context.amplify.pathManager.getBackendDirPath(), 'auth', 'userPoolGroups', 'parameters.json');
+
+    /* eslint-disable */
+    const groupParams = {
+      AuthRoleArn: {
+        'Fn::GetAtt': ['AuthRole', 'Arn'],
+      },
+      UnauthRoleArn: {
+        'Fn::GetAtt': ['UnauthRole', 'Arn'],
+      },
+    };
+    /* eslint-enable */
+
+    fs.outputFileSync(userPoolGroupParams, JSON.stringify(groupParams, null, 4));
+    fs.outputFileSync(userPoolGroupFile, JSON.stringify(userPoolGroupPrecedenceList, null, 4));
+
+    context.amplify.updateamplifyMetaAfterResourceAdd('auth', 'userPoolGroups', {
+      service: 'Cognito-UserPool-Groups',
+      providerPlugin: 'awscloudformation',
+      dependsOn: [
+        {
+          category: 'auth',
+          resourceName,
+          attributes: ['UserPoolId', 'AppClientIDWeb', 'AppClientID', 'IdentityPoolId'],
+        },
+      ],
+    });
+  }
+};
+
+export const addAdminAuth = async (
+  context: any,
+  authResourceName: string,
+  operation: 'update' | 'add',
+  adminGroup?: string,
+  functionName?: string,
+) => {
+  if (adminGroup) {
+    if (!functionName) {
+      const [shortId] = uuid().split('-');
+      functionName = `AdminQueries${shortId}`;
+    }
+    await createAdminAuthFunction(context, authResourceName, functionName, adminGroup, operation);
+    await createAdminAuthAPI(context, authResourceName, functionName, operation);
+  }
+};
+
+export const createAdminAuthFunction = async (
```

Comment:
It would be nice if we could do this by loosely coupling functions category here and let it do its job, instead of duplicating and managing the same logic here... (I know it is out of scope, just wanted to share)

Reply:
This is all copied over from existing code, but I agree 100%. We really need a platform level feature that lets categories make requests to other categories. We're actually getting close to that reality with the headless work because we could expose the headless entry point to other categories through the platform. But yeah, out of scope for now.